build issue

Hello, can't build the code on 4.13.16
Here is the log:



  CC [M]  /home/sergey/Projects/drivers/sunxi/orangepipc/sunxi-mali/r6p2/src/devicedrv/mali/linux/mali_osk_timers.o
/home/sergey/Projects/drivers/sunxi/orangepipc/sunxi-mali/r6p2/src/devicedrv/mali/linux/mali_osk_timers.c: In function ‘_mali_osk_timer_setcallback’:
/home/sergey/Projects/drivers/sunxi/orangepipc/sunxi-mali/r6p2/src/devicedrv/mali/linux/mali_osk_timers.c:63:2: error: implicit declaration of function ‘timer_setup’; did you mean ‘ktime_set’? [-Werror=implicit-function-declaration]
  timer_setup(&tim->timer, callback, 0);
  ^~~~~~~~~~~
  ktime_set
cc1: some warnings being treated as errors
scripts/Makefile.build:302: recipe for target '/home/sergey/Projects/drivers/sunxi/orangepipc/sunxi-mali/r6p2/src/devicedrv/mali/linux/mali_osk_timers.o' failed
make[2]: *** [/home/sergey/Projects/drivers/sunxi/orangepipc/sunxi-mali/r6p2/src/devicedrv/mali/linux/mali_osk_timers.o] Error 1
Makefile:1512: recipe for target '_module_/home/sergey/Projects/drivers/sunxi/orangepipc/sunxi-mali/r6p2/src/devicedrv/mali' failed
make[1]: *** [_module_/home/sergey/Projects/drivers/sunxi/orangepipc/sunxi-mali/r6p2/src/devicedrv/mali] Error 2
make[1]: Leaving directory '/home/sergey/Projects/armbian/build/cache/sources/linux-mainline/linux-4.13.y'

--- a/build.sh
+++ b/build.sh
@@ -14,11 +14,24 @@ apply_patches() {
 
     for patch in $1/*.patch;
     do
-	patch $3 -p1 < $patch
+	patch -p1 < $patch
     done
 
     popd
 }
+
+unapply_patches() {
+    pushd $2
+
+    patches=($1/*.patch)
+    for ((i=${#patches[@]}-1; i>=0; i--));
+    do
+	patch -R -p1 < "${patches[$i]}"
+    done
+
+    popd
+}
+
 
 build_driver() {
     local driver_dir=$(pwd)/$RELEASE/src/devicedrv/mali/
@@ -55,7 +68,7 @@ do
 	    ;;
 	u)
 	    echo "unapplying patches"
-	    apply_patches $(pwd)/patches $RELEASE -R
+	    unapply_patches $(pwd)/patches $RELEASE
 	    ;;
 	\?)
 	    echo "invalid option -$OPTARG"

--- a/patches/0013-mali-support-building-against-4.15.patch
+++ b/patches/0013-mali-support-building-against-4.15.patch
@@ -9,8 +9,8 @@ Signed-off-by: John Stultz <john.stultz@linaro.org>
  src/devicedrv/mali/common/mali_group.c          | 8 +++++---
  src/devicedrv/mali/common/mali_osk_types.h      | 8 +++++++-
  src/devicedrv/mali/linux/mali_memory_os_alloc.c | 2 +-
- src/devicedrv/mali/linux/mali_osk_timers.c      | 8 +-------
- 5 files changed, 15 insertions(+), 13 deletions(-)
+ src/devicedrv/mali/linux/mali_osk_timers.c      | 16 ++++++++++------
+ 5 files changed, 24 insertions(+), 12 deletions(-)
 
 --- a/src/devicedrv/mali/common/mali_control_timer.c
 +++ b/src/devicedrv/mali/common/mali_control_timer.c
@@ -85,14 +85,23 @@ Signed-off-by: John Stultz <john.stultz@linaro.org>
  #if defined(CONFIG_ARM) && !defined(CONFIG_ARM_LPAE)
 --- a/src/devicedrv/mali/linux/mali_osk_timers.c
 +++ b/src/devicedrv/mali/linux/mali_osk_timers.c
-@@ -18,16 +18,11 @@
+@@ -18,16 +18,21 @@
  #include "mali_osk.h"
  #include "mali_kernel_common.h"
  
 -struct _mali_osk_timer_t_struct {
 -	struct timer_list timer;
 -};
--
++#if LINUX_VERSION_CODE < KERNEL_VERSION(4, 14, 0)
++static inline void timer_setup(struct timer_list *timer,
++			       void (*callback)(struct timer_list *),
++			       unsigned int flags)
++{
++	__setup_timer(timer, (TIMER_FUNC_TYPE)callback,
++		      (TIMER_DATA_TYPE)timer, flags);
++}
++#endif
+
  typedef void (*timer_timeout_function_t)(unsigned long);
  
  _mali_osk_timer_t *_mali_osk_timer_init(void)

--- a/patches/0014-mali-Make-devfreq-optional.patch
+++ b/patches/0014-mali-Make-devfreq-optional.patch
@@ -15,10 +15,10 @@ Signed-off-by: Maxime Ripard <maxime.ripard@free-electrons.com>
  r6p2/src/devicedrv/mali/linux/mali_devfreq.c | 7 +++++--
  1 file changed, 5 insertions(+), 2 deletions(-)
 
-diff --git a/r6p2/src/devicedrv/mali/linux/mali_devfreq.c b/r6p2/src/devicedrv/mali/linux/mali_devfreq.c
+diff --git a/src/devicedrv/mali/linux/mali_devfreq.c b/src/devicedrv/mali/linux/mali_devfreq.c
 index 0b0ba1481eed..c3a2636e6635 100755
---- a/r6p2/src/devicedrv/mali/linux/mali_devfreq.c
-+++ b/r6p2/src/devicedrv/mali/linux/mali_devfreq.c
+--- a/src/devicedrv/mali/linux/mali_devfreq.c
++++ b/src/devicedrv/mali/linux/mali_devfreq.c
 @@ -232,8 +232,11 @@ int mali_devfreq_init(struct mali_device *mdev)
  	dp->get_cur_freq = mali_devfreq_cur_freq;
  	dp->exit = mali_devfreq_exit;

